xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxClient.cpp

```diff
@@ -0,0 +1,575 @@
+{
+  Abort();
+
+  m_pInput = pInput;
+  m_IDemux = dynamic_cast<CDVDInputStream::IDemux*>(m_pInput);
+  if (!m_IDemux)
+    return false;
+
+  return m_IDemux->OpenDemux();
+}
+
+void CDVDDemuxClient::Dispose()
+{
+  DisposeStreams();
+
+  m_pInput = nullptr;
+  m_IDemux = nullptr;
+}
+
+void CDVDDemuxClient::DisposeStream(int iStreamId)
+{
+  if (iStreamId < 0 || iStreamId >= MAX_STREAMS)
+    return;
+  delete m_streams[iStreamId];
+  m_streams[iStreamId] = nullptr;
+}
+
+void CDVDDemuxClient::DisposeStreams()
+{
+  for (int i=0; i< MAX_STREAMS; i++)
+    DisposeStream(i);
+}
+
+void CDVDDemuxClient::Reset()
+{
+  CDVDInputStream* pInputStream = m_pInput;
+  Dispose();
+  Open(pInputStream);
+}
+
+void CDVDDemuxClient::Abort()
+{
+  if (m_IDemux)
+    m_IDemux->AbortDemux();
+}
+
+void CDVDDemuxClient::Flush()
+{
+  if (m_IDemux)
+    m_IDemux->FlushDemux();
+}
+
+void CDVDDemuxClient::ParsePacket(DemuxPacket* pkt)
+{
+  CDemuxStream* st = m_streams[pkt->iStreamId];
+  if (st == nullptr)
+    return;
+
+  if (st->ExtraSize)
+    return;
+
+  CDemuxStreamClientInternal* stream = dynamic_cast<CDemuxStreamClientInternal*>(st);
+
+  if (stream == nullptr ||
+     stream->m_parser == nullptr)
+    return;
+
+  if (stream->m_context == nullptr)
+  {
+    AVCodec *codec = avcodec_find_decoder(st->codec);
+    if (codec == nullptr)
+    {
+      CLog::Log(LOGERROR, "%s - can't find decoder", __FUNCTION__);
+      stream->DisposeParser();
+      return;
+    }
+
+    stream->m_context = avcodec_alloc_context3(codec);
+    if (stream->m_context == nullptr)
+    {
+      CLog::Log(LOGERROR, "%s - can't allocate context", __FUNCTION__);
+      stream->DisposeParser();
+      return;
+    }
+    stream->m_context->time_base.num = 1;
+    stream->m_context->time_base.den = DVD_TIME_BASE;
+  }
+
+  if (stream->m_parser_split && stream->m_parser->parser->split)
+  {
+    int len = stream->m_parser->parser->split(stream->m_context, pkt->pData, pkt->iSize);
+    if (len > 0 && len < FF_MAX_EXTRADATA_SIZE)
+    {
+      if (st->ExtraData)
+        delete[] (uint8_t*)st->ExtraData;
+      st->changes++;
+      st->disabled = false;
+      st->ExtraSize = len;
+      st->ExtraData = new uint8_t[len+FF_INPUT_BUFFER_PADDING_SIZE];
+      memcpy(st->ExtraData, pkt->pData, len);
+      memset((uint8_t*)st->ExtraData + len, 0 , FF_INPUT_BUFFER_PADDING_SIZE);
+      stream->m_parser_split = false;
+    }
+  }
+
+
+  uint8_t *outbuf = nullptr;
+  int outbuf_size = 0;
+  int len = av_parser_parse2(stream->m_parser,
+                             stream->m_context, &outbuf, &outbuf_size,
+                             pkt->pData, pkt->iSize,
+                             (int64_t)(pkt->pts * DVD_TIME_BASE),
+                             (int64_t)(pkt->dts * DVD_TIME_BASE),
+                             0);
+  // our parse is setup to parse complete frames, so we don't care about outbufs
+  if(len >= 0)
+  {
+#define CHECK_UPDATE(st, trg, src, invalid) do { \
```

Comment:
Yeah I know ;-) I remember having to debug part of this code once and it was almost impossible due to this macro.